Custom `Promise` implementation no longer works with `assertRejects`

**Describe the bug**

Currently `assertRejects` allows only an instance of `Promise` as return value. If a custom `Promise` is returned, It throws with `AssertionError: Expected function to reject`.

It was working before deno_std@0.140.0.

**Steps to Reproduce**

```ts
class MyPromise implements Promise<void> {
  [Symbol.toStringTag]: string = "Promise";

  then<TResult1 = void, TResult2 = never>(
    onfulfilled?:
      | ((value: void) => TResult1 | PromiseLike<TResult1>)
      | null
      | undefined,
    onrejected?:
      | ((reason: unknown) => TResult2 | PromiseLike<TResult2>)
      | null
      | undefined,
  ): Promise<TResult1 | TResult2> {
    return Promise.reject(new Error("not implemented"))
      .then(onfulfilled)
      .catch(onrejected);
  }

  catch<TResult = never>(
    onrejected?:
      | ((reason: unknown) => TResult | PromiseLike<TResult>)
      | null
      | undefined,
  ): Promise<void | TResult> {
    return Promise.reject(new Error("not implemented")).catch(onrejected);
  }

  finally(onfinally?: (() => void) | null | undefined): Promise<void> {
    return Promise.reject(new Error("not implemented")).finally(onfinally);
  }
}

Deno.test({
  name: "should support custom Promise",
  async fn() {
    await assertRejects(() => new MyPromise());
  },
});
```

Throws with `error: AssertionError: Expected function to reject`.

**Expected behavior**

I would expect the above example to work and `assertRejects` allows a `PromiseLike` object as return value.

```ts
export function assertRejects(
  fn: () => PromiseLike<unknown>,
  msg?: string,
): Promise<unknown>;

Deno.test({
  name: "should support PromiseLike",
  async fn() {
    await assertRejects(() => ({
      then() {
        throw new Error("some error");
      },
    }));
  },
});
```

--- a/testing/README.md
+++ b/testing/README.md
@@ -37,12 +37,12 @@ pretty-printed diff of failing assertion.
   this function does. Also compares any errors thrown to an optional expected
   `Error` class and checks that the error `.message` includes an optional
   string. If there is caught error, it gets returned.
-- `assertRejects()` - Expects the passed `fn` to be async and return a `Promise`
-  that rejects. If the `fn` does not reject, this function will reject _(⚠️ you
-  should normally await this assertion)_. Also optionally accepts an Error class
-  which the expected error must be an instance of, and a string which must be a
-  substring of the error's `.message`. If there is caught error, it gets
-  returned.
+- `assertRejects()` - Expects the passed `fn` to be async and return a
+  `PromiseLike` object that rejects. If the `fn` does not reject, this function
+  will reject _(⚠️ you should normally await this assertion)_. Also optionally
+  accepts an Error class which the expected error must be an instance of, and a
+  string which must be a substring of the error's `.message`. If there is caught
+  error, it gets returned.
 - `unimplemented()` - Use this to stub out methods that will throw when invoked.
 - `unreachable()` - Used to assert unreachable code.
 

--- a/testing/asserts.ts
+++ b/testing/asserts.ts
@@ -680,14 +680,14 @@ export function assertThrows<E extends Error = Error>(
 
 /** Executes a function which returns a promise, expecting it to reject. */
 export function assertRejects(
-  fn: () => Promise<unknown>,
+  fn: () => PromiseLike<unknown>,
   msg?: string,
 ): Promise<unknown>;
 /** Executes a function which returns a promise, expecting it to reject.
  * If it does not, then it throws. An error class and a string that should be
  * included in the error message can also be asserted. */
 export function assertRejects<E extends Error = Error>(
-  fn: () => Promise<unknown>,
+  fn: () => PromiseLike<unknown>,
   // deno-lint-ignore no-explicit-any
   ErrorClass: new (...args: any[]) => E,
   msgIncludes?: string,
@@ -696,12 +696,12 @@ export function assertRejects<E extends Error = Error>(
 /** @deprecated Use assertRejects(fn, msg) instead, which now returns rejected value
  * and you can assert on it. */
 export function assertRejects(
-  fn: () => Promise<unknown>,
+  fn: () => PromiseLike<unknown>,
   errorCallback: (e: Error) => unknown,
   msg?: string,
 ): Promise<Error>;
 export async function assertRejects<E extends Error = Error>(
-  fn: () => Promise<unknown>,
+  fn: () => PromiseLike<unknown>,
   errorClassOrCallbackOrMsg?:
     // deno-lint-ignore no-explicit-any
     | (new (...args: any[]) => E)
@@ -737,7 +737,11 @@ export async function assertRejects<E extends Error = Error>(
   const msgToAppendToError = msg ? `: ${msg}` : ".";
   try {
     const possiblePromise = fn();
-    if (possiblePromise instanceof Promise) {
+    if (
+      possiblePromise &&
+      typeof possiblePromise === "object" &&
+      typeof possiblePromise.then === "function"
+    ) {
       isPromiseReturned = true;
       await possiblePromise;
     }

--- a/testing/asserts_test.ts
+++ b/testing/asserts_test.ts
@@ -857,6 +857,18 @@ Deno.test("assertRejects with non-error value rejected and error callback", asyn
   );
 });
 
+Deno.test("assertRejects with PromiseLike", async () => {
+  await assertRejects(
+    () => ({
+      then() {
+        throw new Error("some error");
+      },
+    }),
+    Error,
+    "some error",
+  );
+});
+
 Deno.test("assertThrows with non-error value thrown and error class", () => {
   assertThrows(
     () => {

